File: src/etch/bindings/java/util/URLSerializer.java
/* $Id$
 *
 * Created by wert on Sep 4, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.util;

import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.ImportExportHelper;
import etch.bindings.java.msg.StructValue;
import etch.bindings.java.msg.Type;
import etch.bindings.java.support.Class2TypeMap;
import etch.bindings.java.support.Validator_string;
import etch.util.URL;

/**
 * An etch serializer for URL.
 */
public class URLSerializer implements ImportExportHelper
{
	private final static String FIELD_NAME = "urlStr";

	/**
	 * Defines custom fields in the value factory so that the importer
	 * can find them.
	 * @param type
	 * @param class2type
	 */
	public static void init( Type type, Class2TypeMap class2type )
	{
		Field field = type.getField( FIELD_NAME );
		class2type.put( URL.class, type );
		type.setComponentType( URL.class );
		type.setImportExportHelper( new URLSerializer( type, field ) );
		type.putValidator( field, Validator_string.get( 0 ) );
		type.lock();
	}

	/**
	 * @param type
	 * @param field
	 */
	public URLSerializer( Type type, Field field )
	{
		this.type = type;
		this.field = field;
	}
	
	private final Type type;
	
	private final Field field;

	public Object importValue( StructValue struct )
	{
		return new URL( (String) struct.get( field ) );
	}

	public StructValue exportValue( Object value )
	{
		StructValue struct = new StructValue( type );
		struct.put( field, ((URL) value).toString() );
		return struct;
	}
}

File: src/etch/bindings/csharp/util/URLSerializer.cs
using System;
using System.Collections.Generic;
using System.Text;
using Etch.Msg;
using Etch.Support;

namespace Etch.Util
{
    public class URLSerializer : ImportExportHelper
    {
        private readonly static String FIELD_NAME = "urlStr";

        public static void Init( XType type, Class2TypeMap class2type )
        {
            Field field = type.GetField( FIELD_NAME );
            class2type.Add( typeof( URL ), type );
            type.SetComponentType( typeof( URL ) );
            type.SetImportExportHelper( new URLSerializer( type, field ) );
            type.PutValidator( field, Validator_string.Get( 0 ) );
            type.Lock();
        }

        public URLSerializer( XType type, Field field )
        {
            this.type = type;
            this.field = field;
        }

        private readonly XType type;
        private readonly Field field;

        public override Object ImportValue( StructValue sv )
	    {
		    return new URL( (String) sv.Get( field ) );
	    }

	    public override StructValue ExportValue( Object value )
	    {
		    StructValue sv = new StructValue( type );
		    sv.Add( field, ((URL) value).ToString() );
		    return sv;
	    }
    }
}
